compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp

```diff
@@ -172,45 +289,275 @@ static void padConvOp(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
     }
   }
+
+  // Extract slice.
+  auto resultType = cast<RankedTensorType>(linalgOp->getResult(0).getType());
+  ArrayRef<int64_t> resultShape = resultType.getShape();
+  int64_t resultRank = resultType.getRank();
+  auto one = rewriter.getIndexAttr(1);
+  SmallVector<OpFoldResult> offsets(resultRank, zero), strides(resultRank, one),
+      sizes;
+  for (auto [dimIdx, dimSize] : llvm::enumerate(resultShape)) {
+    if (ShapedType::isDynamic(dimSize))
+      sizes.push_back(rewriter
+                          .create<tensor::DimOp>(
+                              loc, linalgOp.getDpsInitOperand(0)->get(), dimIdx)
+                          .getResult());
+    else
+      sizes.push_back(rewriter.getIndexAttr(dimSize));
+  }
+  rewriter.replaceOpWithNewOp<tensor::ExtractSliceOp>(linalgOp, paddedCompute,
+                                                      offsets, sizes, strides);
+}
+
+struct PadToIntrinsicsPass
+    : public impl::PadToIntrinsicsBase<PadToIntrinsicsPass> {
+  using Base::Base;
+  void runOnOperation() override;
 };
 
 } // namespace
 
+void PadToIntrinsicsPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  RewritePatternSet patterns(context);
+  auto funcOp = getOperation();
+  bool padConvOps = padTargetType == PadTargetType::ConvOp ||
+                    padTargetType == PadTargetType::All;
+  bool padContractionOps = padTargetType == PadTargetType::ContractionOp ||
+                           padTargetType == PadTargetType::All;
+  SmallVector<linalg::LinalgOp> targetOps;
+  funcOp.walk([&](linalg::LinalgOp linalgOp) {
+    if (isa<linalg::Conv2DNhwcHwcfOp>(linalgOp.getOperation()) && padConvOps) {
+      // Add convOps into worklist.
+      targetOps.push_back(linalgOp);
+    } else if (isa<linalg::BatchMatmulOp, linalg::MatmulOp,
+                   linalg::MatmulTransposeBOp>(linalgOp.getOperation()) &&
+               padContractionOps) {
+      // Add named contractionOps into worklist.
+      targetOps.push_back(linalgOp);
+    } else if (isa<linalg::GenericOp>(linalgOp.getOperation()) &&
+               linalg::isaContractionOpInterface(linalgOp) &&
+               padContractionOps) {
+      // Add named generic contractionOps into worklist.
+      targetOps.push_back(linalgOp);
+    }
+  });
+
+  IRRewriter rewriter(context);
+  for (auto linalgOp : targetOps) {
+    rewriter.setInsertionPoint(linalgOp);
+    TypeSwitch<Operation *, void>(linalgOp.getOperation())
+        .Case<linalg::Conv2DNhwcHwcfOp>(
+            [&](auto convOp) { padConvOp(rewriter, linalgOp); })
+        .Case<linalg::BatchMatmulOp, linalg::MatmulOp,
+              linalg::MatmulTransposeBOp, linalg::GenericOp>(
+            [&](auto matmulOp) { padContractionLikeOp(rewriter, linalgOp); })
+        .Default([&](Operation *op) {
+          // Should not happen since we only add ops we know into the
+          // worklist.
+          assert(false && "Unsupported op to pad.");
```

Comment:
You know what, I think this would actually be more straightforward if we had 2 separate vectors instead of mixing all of these in a single one. The type switch and the assertion would go away

Reply:
Yeah, two separate vectors is what I was looking for as well.

Reply:
That makes sense, I think Mahesh recommended this as well earlier haha

Reply:
I just saw @MaheshRavishankar new reply on this after posting this comment, but yeah, changed/done. :)